driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -212,7 +216,7 @@ public ResponseBuffers receiveMessage(final int responseTo) {
         } else {
             while (!messages.containsKey(responseTo)) {
                 try {
-                    reading.acquire();
+                    readerLock.lock();
```

Comment:
The async reader is now using readerLock in a different way than the sync reader, so they won't work together properly.  

If we end up using this design, I'm not sure yet how to make it work with concurrent sync and async readers.

But we don't have any tests of concurrent async and sync readers.